Generic Element Types

AntonReinhard opened the pull request:
Preparation work for https://github.com/QEDjl-project/QEDcore.jl/issues/60

- Adds `momentum_type` for psp and ps or their types, returning their momentum types (implemented on the interface, no breaking changes)
- Adds `momentum_eltype` for psp and ps or their types, returning their momentum eltypes (implemented on the interface, no breaking changes)
- Make AbstractFourMomentum generic, may have breaking changes.

Fixes #147

AntonReinhard committed:
Add eltype-esque functions, make FourMomentum generic

momentum_type and momentum_eltype for psp and ps, eltype for AbstractFourMomentum

AntonReinhard committed:
Fix #147

AntonReinhard committed:
Fix docs

AntonReinhard committed:
Add tests for momentum_[el]type

CI_INTG_PKG_URL_QEDcore: https://github.com/AntonReinhard/QEDcore.jl#60-generic-eltypes
CI_INTG_PKG_URL_QEDevents: https://github.com/AntonReinhard/QEDevents.jl#eltypes-integ-fix

AntonReinhard committed:
Add type parameter for AbstractGammaRepresentation

CI_INTG_PKG_URL_QEDcore: https://github.com/AntonReinhard/QEDcore.jl#60-generic-eltypes
CI_INTG_PKG_URL_QEDevents: https://github.com/AntonReinhard/QEDevents.jl#eltypes-integ-fix

szabo137 commented on src/implementations/cross_section/diff_probability.jl:
```diff
@@ -21,7 +21,7 @@ Return differential probability evaluated on a phase space point without checkin
     matrix_elements_sq = _matrix_element_square(psp)
 
-    normalization = _averaging_norm(psp.proc)
+    normalization = convert(momentum_eltype(psp), (_averaging_norm(psp.proc)))
```
We should not do `convert` in this *hot* kernel functions. 

I think we should use multiple dispatch here to let Julia select the right function for the passed in element type.

AntonReinhard replied:
I remember we talked about this problem.
Are you suggesting passing to `_averaging_norm` the "requested" type? If so, a thought I just had is to implement this using a `Rational` number, since the norm seems to be a simple fraction anyway. This would preserve the type of whatever it's being used with, as long as we assume the matrix element square won't be `Int` here, which is probably a reasonable assumption.

AntonReinhard replied:
Problem with this idea is that `_averaging_norm` is an interface function and we can't really check that it's actually returning a `Rational`...

szabo137 replied:
The most satisfying solution would be to change the interface and replace `_averaging_norm` with something like `_averaged_number_of_spinpols`, which returns the denominator of `_averaging_norm` instead of the full quotient. This replaces the multiplication with a quotient in the calculation of the differential cross section to just a division by an integer, where the latter should promote automatically to the type of the remaining factors in the cross section. 

I think for now it's fine, but I would like to change this after the generic element types are fully introduced.

AntonReinhard replied:
Take a look at 48f32eb96d731255566547fae53bec7d0a15deef
I can revert it if you think it was nicer before, but this might be a good middle way for now at least.

szabo137 replied:
I think this is a nice middle way, indeed. However, you could consider implementing a more idiomatic version of it: 

```julia
_averaging_norm(::Type{T},proc::MyProcess)
```

instead of using keyword arguments.

AntonReinhard replied:
Fair point, I'll do that

szabo137 commented on src/mocks/cross_section/groundtruth.jl:
```diff
@@ -115,31 +126,49 @@ end
+    in_ps::AbstractMatrix{NTuple{I,<:AbstractMockMomentum}},
+    out_ps::AbstractMatrix{NTuple{O,<:AbstractMockMomentum}},
+) where {I,O}
+    RES_TYPE = eltype(eltype(eltype(in_ps)))
```
Hm, looks weird 😐

AntonReinhard replied:
Well, the `in_ps` is a matrix, the eltype of it is an ntuple, the eltype of that is a four momentum, and the eltype of that is what I want. The only other way I can think of is to type it all out in the `where` statement, but i doubt it'll be much prettier...

szabo137 replied:
understood.

AntonReinhard committed:
JuliaFormatter version compat

CI_INTG_PKG_URL_QEDcore: https://github.com/AntonReinhard/QEDcore.jl#60-generic-eltypes
CI_INTG_PKG_URL_QEDevents: https://github.com/AntonReinhard/QEDevents.jl#eltypes-integ-fix

szabo137 reviewed (dismissed):
LGTM, just two questions regarding possible not-so-good practices.

AntonReinhard committed:
Change _averageing_norm function signature to get correct type

CI_INTG_PKG_URL_QEDcore: https://github.com/AntonReinhard/QEDcore.jl#60-generic-eltypes
CI_INTG_PKG_URL_QEDevents: https://github.com/AntonReinhard/QEDevents.jl#eltypes-integ-fix

AntonReinhard committed:
Change average_norm signature to be more idiomatic

CI_INTG_PKG_URL_QEDcore: https://github.com/AntonReinhard/QEDcore.jl#60-generic-eltypes
CI_INTG_PKG_URL_QEDevents: https://github.com/AntonReinhard/QEDevents.jl#eltypes-integ-fix
CI_INTG_PKG_URL_QEDprocesses: https://github.com/AntonReinhard/QEDprocesses.jl#generic_types
